subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/HealthExpirationStrategy.java

```diff
@@ -16,26 +16,115 @@
+     */
+    public static final String DISABLE_PERFORMANCE_LOGGING = "org.gradle.daemon.performance.disable-logging";
 
-    public HealthExpirationStrategy(DaemonMemoryStatus memoryStatus) {
-        this.strategy = new AnyDaemonExpirationStrategy(ImmutableList.of(
-            new GcThrashingDaemonExpirationStrategy(memoryStatus),
-            new LowHeapSpaceDaemonExpirationStrategy(memoryStatus),
-            new LowNonHeapDaemonExpirationStrategy(memoryStatus)
-        ));
+    /**
+     * The prefix for the message logged when an unhealthy condition is detected.
+     * Used to strip this message from the logs during integration testing.
+     */
+    public static final String EXPIRE_DAEMON_MESSAGE = "The Daemon will expire ";
+
+    private static final Logger LOGGER = Logging.getLogger(HealthExpirationStrategy.class);
+
+    public final DaemonHealthStats stats;
+    public final GarbageCollectorMonitoringStrategy strategy;
+
+    public HealthExpirationStrategy(DaemonHealthStats stats, GarbageCollectorMonitoringStrategy strategy) {
+        this.stats = stats;
+        this.strategy = strategy;
     }
 
     @Override
     public DaemonExpirationResult checkExpiration() {
-        return strategy.checkExpiration();
+        String enabledValue = System.getProperty(ENABLE_PERFORMANCE_MONITORING, "true");
+        if (!Boolean.parseBoolean(enabledValue)) {
+            return DaemonExpirationResult.NOT_TRIGGERED;
+        }
+
+        DaemonExpirationStatus expirationStatus = DO_NOT_EXPIRE;
+        List<String> reasons = new ArrayList<>();
+
+        GarbageCollectionStats heapStats = stats.getHeapStats();
+        if (heapStats.isValid() && heapStats.getEventCount() >= 5
+            && strategy.isAboveHeapUsageThreshold(heapStats.getUsedPercent())
+        ) {
+            if (strategy.isAboveGcThrashingThreshold(heapStats.getGcRate())) {
+                reasons.add("since the JVM garbage collector is thrashing");
+                expirationStatus = highestPriorityOf(IMMEDIATE_EXPIRE, expirationStatus);
+            } else if (strategy.isAboveGcRateThreshold(heapStats.getGcRate())) {
+                reasons.add("after running out of JVM heap space");
+                expirationStatus = highestPriorityOf(GRACEFUL_EXPIRE, expirationStatus);
+            }
+        }
+
+        GarbageCollectionStats nonHeapStats = stats.getNonHeapStats();
+        if (nonHeapStats.isValid() && nonHeapStats.getEventCount() >= 5
+            && strategy.isAboveNonHeapUsageThreshold(nonHeapStats.getUsedPercent())
+        ) {
+            reasons.add("after running out of JVM Metaspace");
+            expirationStatus = highestPriorityOf(GRACEFUL_EXPIRE, expirationStatus);
+        }
+
+        if (expirationStatus == DO_NOT_EXPIRE) {
+            return DaemonExpirationResult.NOT_TRIGGERED;
+        }
+
+        String reason = Joiner.on(" and ").join(reasons);
+        if (!Boolean.getBoolean(DISABLE_PERFORMANCE_LOGGING)) {
```

Comment:
Looking mostly for input here on the text of this message. Also curious if there is a way to perhaps integrate a bit into the `DeprecationLogger` and/or ensure this warning gets into build scans.

Reply:
Added some extra context depending on the expiration type.